transformer_engine/pytorch/module/grouped_linear.py

```diff
@@ -178,17 +178,18 @@ def forward(
 
         if is_grad_enabled:
 
-            saved_inputs, saved_weights = [], []
             ctx.weights_shape_1 = weights[0].shape[1]
 
             tensors_to_save, tensor_objects = prepare_for_saving(*inputmats, *weights_fp8, *biases)
             ctx.save_for_backward(*tensors_to_save)
             ctx.tensor_objects = tensor_objects
 
             ctx.weights_requires_grad = weights[0].requires_grad
+            if fuse_wgrad_accumulation and ctx.weights_requires_grad:
+                ctx.main_grads = [weights[i].main_grad for i in range(num_gemms)]
```

Comment:
It's recommended to use [`ctx.save_for_backward`](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.save_for_backward.html#torch.autograd.function.FunctionCtx.save_for_backward) instead of storing tensors directly in `ctx`. They warn about messing up the grad graph and memory leaks, although I'm not sure what cases they are specifically worried about.

Reply:
I agree. We were saving `main_grad` tensors in `ctx.save_for_backward` in TE 1.x. But I'm see there is comment here. https://github.com/NVIDIA/TransformerEngine/blob/6673f1658c605f7999e04a43bde1493c4b9ad741/transformer_engine/pytorch/module/linear.py#L356
I'm wondering if we have seeing issues with `ctx.save_for_backward`?

Reply:
Interesting, we should follow the example of `Linear` then.

@ksivaman This change is from commit `7e58678` in the internal repo. Do you remember why we can't store `main_grad` in `saved_tensors`?

Reply:
Oh I know, previous `prepare_for_saving` saves `tensor.data` instead of the tensor itself. So for `main_grad` that need to be modified inplace, this could be an issue.

Now #1474 changed `prepare_for_saving` to save the tensor itself, this is no longer a problem.